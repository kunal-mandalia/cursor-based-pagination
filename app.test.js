const axios = require('axios').default;
const { start, stop } = require('./app');

const port = 8081;
const baseEndpoint = `http://0.0.0.0:${port}`;


function createUsersQuery(first = 2, after, sort) {
  const firstInput = `first:${first}`;
  const afterInput = after ? `,after:"${after}"` : '';
  const sortInput = sort ? `,sort: {field:${sort.field},order:${sort.order}}` : '';
  return `
  {
    Users(input: {${firstInput}${afterInput}${sortInput}}) {
      edges {
        node {
          id
          name
          age
          created_at
        }
        cursor
      }
      pageInfo {
        endCursor
      }
    }
  }
`;
}

async function findUsers({ first, after = undefined, sort = undefined }) {
  const query = createUsersQuery(first, after, sort);
  const response = await axios.post(`${baseEndpoint}/graphql`, { query });
  return response.data.data.Users[0];
}

async function wait(ms = 1000) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

beforeAll(async () => {
  await start(port);
});

afterAll(async (done) => {
  await stop();
  await wait();
  done();
});

test('status endpoint reports UP', async () => {
  const { data } = await axios.get(`${baseEndpoint}/status`);
  expect(data).toEqual({ status: 'UP' });
});

test('return correct number of users', async () => {
  const users = await findUsers({ first: 3 });
  expect(users.edges.length).toBe(3);
});

test('users should be ordered by name ascendingly', async () => {
  const input = {
    first: 2,
    sort: { field: 'name', order: 'asc' },
  };
  const users = await findUsers(input);
  expect(users).toMatchSnapshot();
});

test('users should be ordered by name descendingly', async () => {
  const input = {
    first: 2,
    sort: { field: 'name', order: 'desc' },
  };
  const users = await findUsers(input);
  expect(users).toMatchSnapshot();
});
