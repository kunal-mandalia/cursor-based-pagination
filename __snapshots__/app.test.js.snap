// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users should be ordered by name ascendingly 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "1546300800000___name_Alice_asc",
      "node": Object {
        "age": 20,
        "created_at": "1546300800000",
        "name": "Alice",
      },
    },
    Object {
      "cursor": "1551398400000___name_Bob_asc",
      "node": Object {
        "age": 35,
        "created_at": "1551398400000",
        "name": "Bob",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "1551398400000___name_Bob_asc",
  },
}
`;

exports[`users should be ordered by name descendingly 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "1583020800000___name_Zen_desc",
      "node": Object {
        "age": 90,
        "created_at": "1583020800000",
        "name": "Zen",
      },
    },
    Object {
      "cursor": "1527811200000___name_Ed_desc",
      "node": Object {
        "age": 22,
        "created_at": "1527811200000",
        "name": "Ed",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "1527811200000___name_Ed_desc",
  },
}
`;
